From 3234a1f26d0e229761580b6b36c8a40125c47742 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:29:52 -0400
Subject: [net] openvswitch: Use TCP flags in the flow key for stats

Message-id: <7ba9def19925b843ff14514f1cb73698bef72ed4.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94024
O-Subject: [RHEL7.1 net PATCH 19/50] openvswitch: Use TCP flags in the flow key for stats.
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit 88d73f6c411ac2f057829b93b3cf202ee551f6cb
Author: Jarno Rajahalme <jrajahalme@nicira.com>
Date:   Thu Mar 27 12:51:49 2014 -0700

    openvswitch: Use TCP flags in the flow key for stats.

    We already extract the TCP flags for the key, might as well use that
    for stats.

    Signed-off-by: Jarno Rajahalme <jrajahalme@nicira.com>
    Acked-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: Jesse Gross <jesse@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/flow.c b/net/openvswitch/flow.c
index 432f04d..e0fc12b 100644
--- a/net/openvswitch/flow.c
+++ b/net/openvswitch/flow.c
@@ -69,14 +69,12 @@ void ovs_flow_stats_update(struct sw_flow *flow, struct sk_buff *skb)
 
  stats = rcu_dereference(flow->stats[node]);
 
- if ((flow->key.eth.type == htons(ETH_P_IP) ||
-      flow->key.eth.type == htons(ETH_P_IPV6)) &&
-     flow->key.ip.frag != OVS_FRAG_TYPE_LATER &&
-     flow->key.ip.proto == IPPROTO_TCP &&
-     likely(skb->len >= skb_transport_offset(skb) + sizeof(struct tcphdr))) {
-  tcp_flags = TCP_FLAGS_BE16(tcp_hdr(skb));
+ if (likely(flow->key.ip.proto == IPPROTO_TCP)) {
+  if (likely(flow->key.eth.type == htons(ETH_P_IP)))
+   tcp_flags = flow->key.ipv4.tp.flags;
+  else if (likely(flow->key.eth.type == htons(ETH_P_IPV6)))
+   tcp_flags = flow->key.ipv6.tp.flags;
  }
-
  /* Check if already have node-specific stats. */
  if (likely(stats)) {
   spin_lock(&stats->lock);
-- 
1.7.1